jax/experimental/pjit.py

```diff
@@ -108,6 +116,72 @@ def wrapped(*args, **kwargs):
 
   return wrapped
 
+class ParsedPartitionSpec:
+  def __init__(self, user_spec, partitions):
+    self.partitions = tuple(partitions)
+    self.user_spec = user_spec
+
+  @classmethod
+  def from_user_input(cls, entry, arg_name):
+    if entry is None:
+      return cls(entry, ())
+    if not isinstance(entry, PartitionSpec):
```

Comment:
This reminds me that I meant to make pjit accept tuples like xmap does, so we can get rid of PartitionSpecs... :grimacing: This is fine for now though, we change it when we allow tuples.

Reply:
Yeah. This should be really easy after this patch. We just need to update the `is_leaf` function and this parsing method. The rest of the code should adapt just fine, thanks to this early parsing.